addons/im_livechat/static/src/legacy/public_livechat_chatbot.js

```diff
@@ -0,0 +1,874 @@
+    //--------------------------------------------------------------------------
+
+    /**
+     * Add message posted by the bot into the conversation.
+     * This allows not having to wait for the bus (since we run checks based on messages in the
+     * conversation, having the result be there immediately eases the process).
+     *
+     * It also helps while running test tours since those don't have the bus enabled.
+     */
+    _chatbotAddMessage: function (message, options) {
+        message.body = utils.Markup(message.body);
+        this._addMessage(message, options);
+        if (this._chatWindow.isFolded() || !this._chatWindow.isAtBottom()) {
+            this._livechat.incrementUnreadCounter();
+        }
+
+        if (!options || !options.skipRenderMessages) {
+            this._renderMessages();
+        }
+    },
+
+    /**
+     * This method will be transformed into a 'debounced' version (see init).
+     *
+     * The purpose is to handle steps of type 'free_input_multi', that will let the user type in
+     * multiple lines of text before the bot goes to the next step.
+     *
+     * Every time a 'keydown' is detected into the input, or every time a message is sent, we call
+     * this debounced method, which will give the user about 10 seconds to type more text before
+     * the next step is triggered.
+     *
+     * First we check if the last message was sent by the user, to make sure we always let him type
+     * at least one message before moving on.
+     */
+    _chatbotAwaitUserInput: function () {
+        const lastMessage = this._messages.length !== 0 ? this._messages[this._messages.length - 1] : null;
+        if (lastMessage && lastMessage.getAuthorID() !== this._livechat.getOperatorPID()[0]) {
+            if (!this._chatbotShouldEndScript()) {
+                this._chatbotSetIsTyping();
+                this.nextStepTimeout = setTimeout(
+                    this._chatbotTriggerNextStep.bind(this), this._chatbotMessageDelay);
+            } else {
+                this._chatbotEndScript();
+            }
+        }
+    },
+
+    /**
+     * When the user first interacts with the bot, we want to make sure to actually post the welcome
+     * messages into the conversation.
+     *
+     * Indeed, before that, they are 'virtual' messages that are not tied to mail.messages, see
+     * #_sendWelcomeChatbotMessage() for more information.
+     *
+     * Posting them as real messages allows to have a cleaner model and conversation, that will be
+     * kept intact when changing page on the website.
+     *
+     * It also allows tying any first response / question_selection choice to a chatbot.message
+     * that has a linked mail.message.
+     */
+    _chatbotCheckWelcomeMessages: async function () {
+        const welcomeMessages = this._messages.filter((message) => {
+            return message._id && typeof message._id === 'string' && message._id.startsWith('_welcome_');
+        });
```

Comment:
It will be prettier if you create a separate variable imo
```suggestion
        const welcomeMessages = this._messages.filter(message => message.isWelcome);
```

Reply:
(Same answer as previously).